test_conformance/profiling/profiling_timebase.cpp

```diff
@@ -0,0 +1,104 @@
+
+    // If CL_PLATFORM_HOST_TIMER_RESOLUTION returns 0, clGetDeviceAndHostTimer
+    // is not a supported feature
+    if (timer_resolution == 0 && version >= Version(3, 0))
+    {
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    cl_ulong hostTime;
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+    clEventWrapper kEvent;
+
+    clEventWrapper uEvent = clCreateUserEvent(context, &err);
+    test_error(err, "Failed to create user event");
+
+    err = create_single_kernel_helper(context, &program, &kernel, 1,
+                                      &kernelCode, "kernel_empty");
+    test_error(err, "Failed to create kernel");
+
+    cl_ulong deviceTimeBeforeQueue;
+    err = clGetDeviceAndHostTimer(device, &deviceTimeBeforeQueue, &hostTime);
+    test_error(err, "Unable to get starting device time");
+
+    clEventWrapper userEvents = uEvent;
+    err = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, NULL, NULL, 1,
+                                 &userEvents, &kEvent);
+    test_error(err, "clEnqueueNDRangeKernel failed");
+
+    cl_ulong deviceTimeAfterQueue;
+    err = clGetDeviceAndHostTimer(device, &deviceTimeAfterQueue, &hostTime);
+    test_error(err, "Unable to get queue device time");
+
+    err = clFlush(queue);
+    test_error(err, "clFlush failed");
+
+    err = clSetUserEventStatus(uEvent, CL_COMPLETE);
+    test_error(err, "Unable to complete user event");
+
+    err = clWaitForEvents(1, &kEvent);
+    test_error(err, "clWaitForEvents failed");
+
+    cl_ulong deviceTimeAfterCompletion;
+    err =
+        clGetDeviceAndHostTimer(device, &deviceTimeAfterCompletion, &hostTime);
+    test_error(err, "Unable to get finishing device time");
+
+    cl_ulong eventQueue, eventSubmit, eventStart, eventEnd;
+    err = clGetEventProfilingInfo(kEvent, CL_PROFILING_COMMAND_QUEUED,
+                                  sizeof(cl_ulong), &eventQueue, NULL);
+    test_error(err, "clGetEventProfilingInfo failed");
+    err = clGetEventProfilingInfo(kEvent, CL_PROFILING_COMMAND_SUBMIT,
+                                  sizeof(cl_ulong), &eventSubmit, NULL);
+    test_error(err, "clGetEventProfilingInfo failed");
+    err = clGetEventProfilingInfo(kEvent, CL_PROFILING_COMMAND_START,
+                                  sizeof(cl_ulong), &eventStart, NULL);
+    test_error(err, "clGetEventProfilingInfo failed");
+    err = clGetEventProfilingInfo(kEvent, CL_PROFILING_COMMAND_END,
+                                  sizeof(cl_ulong), &eventEnd, NULL);
+    test_error(err, "clGetEventProfilingInfo failed");
+
+    test_assert_error(deviceTimeBeforeQueue < eventQueue,
+                      "Device timestamp was taken before kernel was queued");
+    test_assert_error(eventQueue < deviceTimeAfterQueue,
+                      "Device timestamp was taken after kernel was queued");
+    test_assert_error(deviceTimeAfterQueue < eventSubmit,
+                      "Device timestamp was taken before kernel was submitted");
```

Comment:
I don't think this is a safe check.  The OpenCL execution model says:

> Submitted: The command is flushed from the command-queue and submitted for execution on the device. Once flushed from the command-queue, a command will execute after any prerequisites for execution are met.

I believe this means that the transition to the SUBMITTED state can happen anytime after the QUEUED state, even if the event dependencies defining the "prerequisites for execution" are not yet met.

Reply:
Agree, this is how I interpreted the quoted paragraph from the specification as well. Assert removed.